src/ra_utility.h

```diff
@@ -172,6 +177,146 @@ _NODISCARD inline auto __cdecl tstrlen(_In_z_ LPCTSTR _Str) noexcept
+    // too difficult as a generic
+    if constexpr(std::is_same_v<CharT, char>)
+    {
+        assert(std::string{ format }.find("%n") == std::string::npos);
+        std::string test{ format };
+        const auto num_specs{ std::count(test.begin(), test.end(), '%') };
+        assert(num_specs == sizeof...(args));
+    }
+    else if constexpr(std::is_same_v<CharT, wchar_t>)
+    {
+        assert(std::wstring{ format }.find(L"%n") == std::wstring::npos);
+        std::wstring test{ format };
+        const auto num_specs{ std::count(test.begin(), test.end(), L'%') };
+        assert(num_specs == sizeof...(args));
+    }
+#else
+    if ((buffer == nullptr) || (format == nullptr))
+        throw std::invalid_argument{ "nullptr arguments are undefined for this operation!" };
+    else if ((bufsz == 0) || (bufsz > RSIZE_MAX))
+    {
+        std::string msg;
+        {
+            std::ostringstream oss;
+            oss << "Invalid buffer size!\nValid range: 0 < bufsz <= " << RSIZE_MAX << '!';
+            msg = oss.str();
+        }
+        throw std::out_of_range{ msg };
+    }
+    else if constexpr (std::is_same_v<CharT, char>)
+    {
+        if (std::string{ format }.find("%n") != std::string::npos)
+            throw std::invalid_argument{ "The conversion specifier '%n' is unsupported!" };
+        {
+            std::string test{ format };
+            const auto num_specs{ std::count(test.begin(), test.end(), '%') };
+            if(num_specs != sizeof...(args))
+                throw std::logic_error{ "The number of format specifiers and arguments must be identical!" };
+        }
+    }
+    else if constexpr (std::is_same_v<CharT, wchar_t>)
+    {
+        if(std::wstring{ format }.find(L"%n") != std::wstring::npos)
+            throw std::invalid_argument{ "The conversion specifier '%n' is unsupported!" };
+        {
+            std::wstring test{ format };
+            const auto num_specs{ std::count(test.begin(), test.end(), L'%') };
+            if (num_specs != sizeof...(args))
+                throw std::logic_error{ "The number of format specifiers and arguments must be identical!" };
+        }
+    }
+#endif /* _DEBUG */
+    
+    // MSVC doesn't seem to have constraint handlers like C11 we could throw an exception...
+    if constexpr (std::is_same_v<CharT, char>)
+    {
+        const auto ret{ std::snprintf(buffer, bufsz, format, std::forward<Args>(args)...) };
+#if _DEBUG
+        assert(std::strlen(buffer) < bufsz);
+        assert(ret > 0);
+#else
+        if (std::strlen(buffer) > bufsz)
+            throw std::out_of_range{ "Stack or heap corruption detected, increase bufsz!" };
+        else if (ret < 0)
+            throw std::runtime_error{ "An encoding error has occurred, try using Unicode!" };
+#endif /* _DEBUG */
+
+        return ret;
+    }
+    if constexpr (std::is_same_v<CharT, wchar_t>)
+    {
+        const auto ret{ std::swprintf(buffer, bufsz, format, std::forward<Args>(args)...) };
```

Comment:
why go through all this effort if it just boils down to changing `wsprintf` to `swprintf`?

Reply:
It's for flexibility so it doesn't strictly have to be TCHAR since there's explicit use of unicode APIs even in a mbcs build. It can be used as a replacement for sprintf in general but didn't want to change all of them at this time.